pkgs/io_file/lib/src/file_system.dart

```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// An abstract representation of a file system.
+base class FileSystem {
+  /// Renames, and possibly moves a file system object from one path to another.
+  ///
+  /// If `newPath` is a relative path, it is resolved against the current
+  /// working directory. This means that simply changing the name of a file,
+  /// but keeping it the original directory, requires creating a new complete
+  /// path with the new name at the end.
+  ///
+  /// TODO: add an example here.
+  ///
+  /// On some platforms, a rename operation cannot move a file between
```

Comment:
Orthogonal to this PR - will there be a capability detection system? Will people be able to query which features a given filesystem object supports?